src/PrestaShopBundle/Form/Admin/ShopParameters/CustomerPreferences/CustomerPreferencesFormHandler.php

```diff
@@ -0,0 +1,158 @@
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2018 PrestaShop SA
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShopBundle\Form\Admin\ShopParameters\CustomerPreferences;
+
+use PrestaShop\PrestaShop\Core\Form\FormDataProviderInterface;
+use PrestaShop\PrestaShop\Core\Form\FormHandlerInterface;
+use PrestaShopBundle\Entity\Repository\TabRepository;
+use PrestaShopBundle\Service\Hook\HookDispatcher;
+use Symfony\Component\Form\FormFactoryInterface;
+use Symfony\Component\Translation\TranslatorInterface;
+
+/**
+ * Class manages "Configure > Shop Parameters > Customer Settings" page
+ * form handling.
+ */
+class CustomerPreferencesFormHandler implements FormHandlerInterface
+{
+    /**
+     * @var FormFactoryInterface
+     */
+    private $formFactory;
+
+    /**
+     * @var FormDataProviderInterface
+     */
+    private $dataProvider;
+
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    /**
+     * @var HookDispatcher
+     */
+    private $dispatcher;
+
+    /**
+     * @var TabRepository
+     */
+    private $tabRepository;
+
+    public function __construct(
+        FormFactoryInterface $formFactory,
+        FormDataProviderInterface $dataProvider,
+        TranslatorInterface $translator,
+        HookDispatcher $dispatcher,
+        TabRepository $tabRepository
+    ) {
+        $this->formFactory = $formFactory;
+        $this->dataProvider = $dataProvider;
+        $this->translator = $translator;
+        $this->dispatcher = $dispatcher;
+        $this->tabRepository = $tabRepository;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getForm()
+    {
+        $builder = $this->formFactory->createBuilder()
+            ->add('general', GeneralType::class)
+            ->setData($this->dataProvider->getData());
+
+        $this->dispatcher->dispatchForParameters('displayCustomerPreferencesForm', [
+            'form_builder' => $builder,
+        ]);
+
+        return $builder->getForm();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function save(array $data)
+    {
+        if ($errors = $this->validate($data)) {
+            return $errors;
+        }
+
+        if (!$errors = $this->dataProvider->setData($data)) {
+            $this->handleB2bUpdate($data['general']['enable_b2b_mode']);
+        }
+
+        $this->dispatcher->dispatchForParameters('actionCustomerPreferencesSave', [
+            'errors' => &$errors,
+            'form_data' => &$data,
+        ]);
+
+        return $errors;
+    }
+
+    /**
+     * Perform validations on form data
+     *
+     * @param array $data
+     *
+     * @return array    Array of errors if any
+     */
+    protected function validate(array $data)
```

Comment:
this validation should be done in FormDataProvider

Reply:
yes! even better :)

Reply:
now that I think about it, I'll use your suggestion and move it to FormDataProvider, to avoid mixed content returned by validateConfiguration() as it should return booleans only.

Reply:
yup, it's difficult to figure out the purpose of what needs to be returned by interfaces in PHP 5. Can't wait to increase our PHP minimum version to PHP 7.1 :+1: